commodore/cluster.py

```diff
@@ -13,14 +13,79 @@
 )
 
 from .config import Config
-
-
-def fetch_cluster(cfg, clusterid):
-    cluster = lieutenant_query(cfg.api_url, cfg.api_token, "clusters", clusterid)
-    # TODO: move Commodore global defaults repo name into Lieutenant
-    # API/cluster facts
-    cluster["base_config"] = "commodore-defaults"
-    return cluster
+from .git import HEAD
+
+
+class Cluster:
+    def __init__(
+        self,
+        cfg: Config,
+        cluster_response: Dict,
+        tenant_response: Dict,
+    ):
+        self.__cfg = cfg
+        self.__cluster = cluster_response
+        self.__tenant = tenant_response
+        if (
+            "tenant" not in self.__cluster
+            or self.__cluster["tenant"] != self.__tenant["id"]
+        ):
+            raise click.ClickException("Customer id mismatch")
+
+    def id(self) -> str:
+        return self.__cluster["id"]
+
+    def global_git_repo_url(self) -> str:
+        field = "globalGitRepoURL"
+        if field not in self.__tenant:
+            return f"{self.__cfg.global_git_base}/commodore-defaults.git"
+        return self.__tenant[field]
+
+    def global_git_repo_revision(self) -> str:
+        field = "globalGitRepoRevision"
+        return self.__cluster.get(field, self.__tenant.get(field, HEAD))
+
+    def config_repo_url(self) -> str:
+        repo_url = self.__tenant.get("gitRepo", {}).get("url", None)
+        if repo_url is None:
+            raise click.ClickException(
+                " > API did not return a repository URL for tenant '%s'"
+                % self.__cluster["tenant"]
+            )
+        return repo_url
+
+    def config_git_repo_revision(self) -> str:
+        field = "gitRepoRevision"
+        return self.__cluster.get(field, self.__tenant.get(field, HEAD))
+
+    def catalog_repo_url(self) -> str:
+        repo_url = self.__cluster.get("gitRepo", {}).get("url", None)
+        if repo_url is None:
+            raise click.ClickException(
+                " > API did not return a repository URL for cluster '%s'"
+                % self.__cluster["id"]
+            )
+        return repo_url
+
+    def tenant(self) -> str:
+        return self.__tenant["id"]
+
+    def facts(self) -> Dict[str, str]:
+        if "facts" not in self.__cluster:
+            return {}
```

Comment:
Is it allowed that a cluster API response doesn't have the field facts? Or should this be an error?

Reply:
Conceptually, it would be OK to not have facts. On the other hand, we do have an assertion somewhere else expecting the `cloud` and `distribution` facts. This has partially become obsolete. Once we remove the backwards compatibility layer, Commodore does no longer need any facts. It is up the the hierarchy defined in the global config to define if and which facts are required.